### PR TITLE
Fix Dynamo .grad reads for new in-graph parameters

### LINKED ISSUE #171204: Dynamo can't trace a code when we construct nn.Parameter in the forward.
### 🐛 Describe the bug

```python 
import torch
import torch._dynamo

torch._dynamo.config.graph_break_on_nn_param_ctor = False

def fn(x):
    w = torch.nn.Parameter(torch.ones(4, 4))
    if w.grad is None:
        w.grad = torch.zeros_like(w)
    return w.grad + x

x = torch.randn(4, 4)
compiled_fn = torch.compile(fn, backend='eager', fullgraph=True)
result = compiled_fn(x)
```
```
Unsupported: Failed to trace builtin operator
  Explanation: Dynamo does not know how to trace builtin operator `add` with argument types ['<unknown type>', 'Tensor'] (has_kwargs False)
  Hint: Avoid calling builtin `add` with argument types ['<unknown type>', 'Tensor']. Consider using an equivalent alternative function/method to `add`.
  Hint: If you are attempting to call a logging function (e.g. `print`), you can try adding it to `torch._dynamo.config.reorderable_logging_functions`.
  Hint: Please report an issue to PyTorch.

  Developer debug context: builtin add [<class 'torch._dynamo.variables.misc.GetAttrVariable'>, <class 'torch._dynamo.variables.tensor.TensorVariable'>] False

 For more details about this graph break, please visit: https://meta-pytorch.github.io/compile-graph-break-site/gb/gb0059.html

from user code:
   File "/tmp/ipykernel_616085/151731544.py", line 10, in fn
    return w.grad + x

Set TORCHDYNAMO_VERBOSE=1 for the internal stack trace (please do this especially if you're reporting a bug to PyTorch). For even more developer context, set TORCH_LOGS="+dynamo"
```
I think this is due to the fact we are emitting generic GetAttr node for w.grad instead of proper type. 

### Versions

main

cc @ezyang @gchanan @kadeng @msaroufim @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chauhang @amjames @Lucaskabela @jataylo

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #184972

Dynamo represents tensors created inside a compiled region with tracked
attribute mutation state. The first read of `.grad` on such a newly created
leaf tensor should observe eager's initial `None` value, but the existing
handler only returned a concrete value after a pending `.grad` mutation was
already tracked. The initial read therefore fell through to the generic getattr
path, so `w.grad is None` traced incorrectly and later tensor arithmetic saw a
GetAttrVariable instead of the assigned grad tensor.

Fix this at the source by returning a constant `None` for newly created leaf
tensors before any pending `.grad` mutation exists. Sourced tensors continue to
use the existing source-aware fallback, which avoids the broader `.grad` wrapper
approach from the abandoned PR and preserves non-input parameter grad handling.

Add focused coverage for constructing an `nn.Parameter` in forward, checking
that the initial `.grad` read is `None`, assigning a grad tensor, and using it
in compiled tensor arithmetic.

Fixes #171204
Generated by my agent

Test Plan:
- python test/dynamo/test_misc.py MiscTests.test_param_grad_in_forward
- python test/dynamo/test_misc.py MiscTests.test_source_non_input_grad_access
- python test/dynamo/test_dynamic_shapes.py DynamicShapesMiscTests.test_source_non_input_grad_access_dynamic_shapes
- python test/dynamo/test_dynamic_shapes.py DynamicShapesMiscTests.test_param_grad_in_forward_dynamic_shapes
- python test/dynamo/test_misc.py MiscTests.test_intermediary_tensor_grad_access
- python test/dynamo/test_misc.py MiscTests.test_grad
- python test/dynamo/test_fwd_loss_bwd.py TestForwardLossBackward.test_autograd_grad_with_ingraph_nn_parameter
- python test/dynamo/test_fwd_loss_bwd.py TestForwardLossBackward.test_tensor_backward_with_ingraph_nn_parameter
- python test/inductor/test_compiled_autograd.py TestCompiledAutograd.test_accumulate_grad_polyfill_case_2_3_1
- python test/inductor/test_compiled_autograd.py TestCompiledAutograd.test_accumulate_grad_polyfill_case_3_2
- python test/dynamo/test_misc.py MiscTests.test_param_grad_in_forward MiscTests.test_source_non_input_grad_access MiscTests.test_intermediary_tensor_grad_access MiscTests.test_grad
- git diff --cached --check
- lintrunner -a

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98 @mlazos

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased and resubmitted the PR to refresh CI after the previous new failures were checkout/submodule credential failures rather than test failures. I also added the `release notes: dynamo` label and reran `lintrunner -a`, `MiscTests.test_param_grad_in_forward`, and `MiscTests.test_source_non_input_grad_access` locally.

File: test/dynamo/test_misc.py [hunk 1]
```diff
@@ -6185,6 +6185,27 @@ def fn(a, b):
...
+    @torch._dynamo.config.patch(graph_break_on_nn_param_ctor=False)
+    def test_param_grad_in_forward(self):
+        cnts = torch._dynamo.testing.CompileCounter()
+
+        def fn(x):
+            w = torch.nn.Parameter(torch.ones(4, 4))
+            saw_none = w.grad is None
+            if saw_none:
+                w.grad = torch.zeros_like(w)
+            return saw_none, w.grad + x
```
Comment (zou3519) [inline test/dynamo/test_misc.py:6195]: can you add a test that does torch.autograd.grad on w inside of fn and then confirm that fn.grad is not None after that? This test can be with fullgraph=False, I just want to check no silent incorrectness

**Review by zou3519 — CHANGES_REQUESTED** [review]
seems fine, just please add a test for the autograd.grad case

File: test/dynamo/test_misc.py (hunk 1 above)
Comment (jansel) [reply test/dynamo/test_misc.py:6195]: My agent says I added `MiscTests.test_param_autograd_grad_in_forward`, which exercises `torch.autograd.grad` on an `nn.Parameter` created inside the compiled function with `fullgraph=False`, checks the returned grad is not `None`, verifies `w.grad` remains `None` like eager, and compares the grad tensor against eager output.

**Review by zou3519 — APPROVED** [review]

**Comment by jansel** [comment]
@claude are those CI failures related to this PR?
